### PR TITLE
Coherence negative impls implied bounds

### LINKED ISSUE #93875: coherence doesn't permit negative impls to take advantage of implied bounds
https://github.com/rust-lang/rust/pull/93652 implemented the "negative coherence" rules, but it left one part undone. When we try to prove `T: !Trait` for some trait, we ought to include in the environment the "wf" bounds for the input types on the impl; the current logic does not. As a result, tests like `src/test/ui/coherence/coherence-negative-outlives-lifetimes.rs` do not compile:

```rust
trait MyPredicate<'a> {}
impl<'a, T> !MyPredicate<'a> for &T where T: 'a {}
trait MyTrait<'a> {}
impl<'a, T: MyPredicate<'a>> MyTrait<'a> for T {}
impl<'a, T> MyTrait<'a> for &'a T {}
```

Here, for `MyTrait` to be legal, we would have to be able to prove that `T: !MyPredicate<'a>`. There is a negative impl, but it requires that `T: 'a`, which we are not able to prove. *However,* we ought to be able to, because the impl in question is for `&'a T`, and that includes a WF bound that `T: 'a`.

cc @spastorino

### PR DESCRIPTION
**Pull request by spastorino**
Fixes #93875

This PR is rebased on top of #100789 and it would need to include that one which is already r+ed.

r? @nikomatsakis 

cc @lcnr (which I've talked about https://github.com/rust-lang/rust/commit/3222f420d9d2312efe0735eb48160c7b070adc54, I guess after you finish your reordering of modules and work with OutlivesEnvironmentEnv this commit can just be reverted).

### CODE REVIEW

**Comment by spastorino** [comment]
Rebased on top of #99702 which has removed `OutlivesEnvironmentExt`.

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 1]
```diff
@@ -380,22 +380,34 @@ fn negative_impl_exists<'cx, 'tcx>(
...
 
-    let errors = super::fully_solve_obligation(infcx, o);
-    if !errors.is_empty() {
+    let param_env = o.param_env;
+    if !super::fully_solve_obligation(&infcx, o).is_empty() {
         return false;
     }
 
-    let outlives_env = OutlivesEnvironment::new(param_env);
+    let outlives_env = if let Some(body_def_id) = body_def_id.as_local() {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:398]: please do not change the used implied bounds based on whether the impl is from the current crate, at least not without a comment explaining the reasoning why.

You can use the `CRATE_HIR_ID` for obligations from foreign crates

Comment (spastorino) [reply compiler/rustc_trait_selection/src/traits/coherence.rs]: @lcnr fixed in https://github.com/rust-lang/rust/pull/100888/commits/4da14ef50eb697298f36d41f91cb02dc800275bf

**Comment by spastorino** [comment]
Rebased on top of #100789, so this is now ready.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
Looks like lcnr already reviewed, but r+ from me, too.

**Pull request merged**